Add odo installation docs (#4923)
openshift/odo: website/docs/getting-started/installation.md

title: Installation
sidebar_position: 4
---

odo can be used as a CLI tool and as an IDE plugin; it can be run on Linux, Windows and Mac systems.

## CLI Binary installation
odo supports amd64 architecture for Linux, Mac and Windows.
Additionally, it also supports amd64, arm64, s390x, and ppc64le architectures for Linux.

> See the [release page](https://mirror.openshift.com/pub/openshift-v4/clients/odo/latest/) for more information.

Reviewer: ```suggestion
See the [release page](https://developers.redhat.com/content-gateway/rest/mirror/pub/openshift-v4/clients/odo/latest/) for more information.
```

### Installing odo on Linux/Mac
```shell
OS="$(uname | tr '[:upper:]' '[:lower:]')"

> ARCH="$(uname -m | sed -e 's/x86_64/amd64/' -e 's/\(arm\)\(64\)\?.*/\1\2/' -e 's/aarch64$/arm64/')"

Author: We are using `uname -m` in some of the scripts in odo, so I think it's safe to say that it works on IBM machines.
See - https://github.com/openshift/odo/blob/main/scripts/openshiftci-presubmit-all-tests.sh#L8
https://github.com/openshift/odo/blob/main/scripts/generate-coverage.sh#L7

> curl -L https://mirror.openshift.com/pub/openshift-v4/clients/odo/latest/odo-$OS-$ARCH -o odo

Reviewer: ```suggestion
curl -L https://developers.redhat.com/content-gateway/rest/mirror/pub/openshift-v4/clients/odo/latest/odo-$OS-$ARCH -o odo
```

sudo install odo /usr/local/bin/
```

### Installing odo on Windows

> 1. Download the [odo-windows-amd64.exe](https://mirror.openshift.com/pub/openshift-v4/clients/odo/latest/odo-windows-amd64.exe) file.

Reviewer: ```suggestion
1. Download the [odo-windows-amd64.exe](https://developers.redhat.com/content-gateway/rest/mirror/pub/openshift-v4/clients/odo/latest/odo-windows-amd64.exe) file.
```

> 2. Rename the downloaded file to odo.exe and move it to a folder of choice, for example C:\odo

Reviewer: ```suggestion
2. Rename the downloaded file to odo.exe and move it to a folder of choice, for example `C:\odo`.
```

3. Add the location of odo.exe to `%PATH%` variable (refer to the steps below).

#### Setting the PATH variable in Windows 10
1. Click **Search** and type `env` or `environment`.
2. Select **Edit environment variables for your account**.
3. Select **Path** from the **Variable** section and click **Edit**.
4. Click **New**, add the location where you copied the odo binary (e.g. `C:\odo` in [Step 2 of Installation](#installing-odo-on-windows) into the field or click **Browse** and select the directory, and click **OK**.

#### Setting the PATH variable in Windows 7/8

> 1. Click **Start** and in the Search box types `Advance System Settings`.

Reviewer: ```suggestion
1. Click **Start** and in the Search box type `Advanced System Settings`.
```

> 2. Select **Advanced systems settings** and click the **Environment Variables** button at the bottom.

Reviewer: ```suggestion
2. Select **Advanced Systems Settings** and click the **Environment Variables** button at the bottom.
```

Author: It seems to be correct. See - https://www.groovypost.com/howto/set-path-system-variable-windows-7-8/

> 3. Select the **Path** variable from the **System variable** section and click **Edit**.

Reviewer: Is it really "System variable"? I don't know, so I'm asking. But I think it might be "System Variables".

Author: I am not sure. I copy-pasted this from the old website.

Author: It is `System variables`. See the 4th image here  https://www.groovypost.com/howto/set-path-system-variable-windows-7-8/.

title: Installation
sidebar_position: 4
---

odo can be used as a CLI tool and as an IDE plugin; it can be run on Linux, Windows and Mac systems.

> ## CLI Binary installation

Author: Right, I like this idea, but I also like @feloy's idea - https://github.com/openshift/odo/pull/4923#discussion_r674952235. What do you think about it @dharmit?

> ## CLI Binary installation

Reviewer: I like it. It's geekier. 👍🏾

### Installing odo on Linux
```shell

>   $ curl -L https://mirror.openshift.com/pub/openshift-v4/clients/odo/latest/odo-linux-amd64 -o /usr/local/bin/odo

Reviewer 2: I think we can auto-detect the system and machine type with `uname`, and get only one instruction for all Linuxes and mac.

An example from https://krew.sigs.k8s.io/docs/user-guide/setup/install/:

```
OS="$(uname | tr '[:upper:]' '[:lower:]')"
ARCH="$(uname -m | sed -e 's/x86_64/amd64/' -e 's/\(arm\)\(64\)\?.*/\1\2/' -e 's/aarch64$/arm64/')"
```

Also, non root user will get an error on trying to write the downloaded file into `/usr/local/bin`. You could instead run (`install` will position the `x` flag correctly):

```
curl ... -o odo
sudo install odo /usr/local/bin/
```

  $ chmod +x /usr/local/bin/odo
```

###  Installing odo on Linux on IBM Power
```shell
  $ curl -L https://mirror.openshift.com/pub/openshift-v4/clients/odo/latest/odo-linux-ppc64le -o /usr/local/bin/odo
  $ chmod +x /usr/local/bin/odo
```

### Installing odo on Linux on IBM Z
```shell
  $ curl -L https://mirror.openshift.com/pub/openshift-v4/clients/odo/latest/odo-linux-s390x -o /usr/local/bin/odo
  $ chmod +x /usr/local/bin/odo
```

### Installing odo on macOS
```shell
  $ curl -L https://mirror.openshift.com/pub/openshift-v4/clients/odo/latest/odo-darwin-amd64 -o /usr/local/bin/odo
  $ chmod +x /usr/local/bin/odo
```

### Installing odo on Windows

> 1. Download the latest odo-windows-amd64.exe file.

Reviewer 2: Download from where?

2. Rename the downloaded file to odo.exe and move it to a folder of choice, for example C:\odo

> 3. Add the location of odo.exe to %PATH%.

Reviewer: ```suggestion
3. Add the location of odo.exe to `%PATH%` variable (refer steps provided below).
```

#### Setting the PATH variable in Windows 10
1. Click **Search** and type `env` or `environment`.
2. Select **Edit environment variables for your account**.
3. Select **Path** from the **Variable** section and click **Edit**.

> 4. Click **New** and type `<location_of_odo_binary>` into the field or click **Browse** and select the directory, and click **OK**.

Reviewer: ```suggestion
4. Click **New**, add the location where you copied the odo binary (e.g. `C:\odo` in [Step 2 of Installation](./installation#installing-odo-on-windows) into the field or click **Browse** and select the directory, and click **OK**.
```

#### Setting the PATH variable in Windows 7/8
1. Click **Start** and in the Search box types `Advance System Settings`.
2. Select **Advanced systems settings** and click the **Environment Variables** button at the bottom.
3. Select the **Path** variable from the **System variable** section and click **Edit**.

> 4. Scroll to the end of the **Variable Value** and add `;<location_of_odo_binary>` and click **OK**.

Reviewer: ```suggestion
4. Scroll to the end of the **Variable Value**, add the location where you copied the odo binary (e.g. `C:\odo` in [Step 2 of Installation](./installation#installing-odo-on-windows) and click **OK**.
```

5. Click **OK** to close the **Environment Variable** dialog.
6. Click **OK** to close the **Systems Properties** dialog.

## Installing odo in Visual Studio Code (VSCode)

> The OpenShift VSCode extension uses both odo and the oc binary to interact with Kubernetes or OpenShift cluster.

Reviewer 2: This does not seem clear to me. Does it use `oc` when running a Kubernetes cluster?

Author: I am not sure, it doesn't mention anything on the [extension homepage](https://marketplace.visualstudio.com/items?itemName=redhat.vscode-openshift-connector).
But I guess you can still use `oc` for your Kubernetes cluster.

1. Open VS Code.
2. Launch VS Code Quick Open (Ctrl+P)
3. Paste the following command:
    ```shell
     $ ext install redhat.vscode-openshift-connector
    ```

> ## Installation from source

Reviewer 2: ```suggestion
## Installing from source
```

> 1. Clone the repository and cd.

Reviewer 2: ```suggestion
1. Clone the repository and cd into it.
```

> ```shell

Reviewer: Every code block in this ordered list needs to have 4 blank spaces or a tab so that it renders properly. You've already ensured that in `Installing odo in Visual Studio Code (VSCode)` section in point 3.

> $ git clone https://github.com/openshift/odo.git; cd odo

Reviewer 2: nitpick: 
```suggestion
$ git clone https://github.com/openshift/odo.git && cd odo
```

Reviewer: Or, put it into separate lines. But don't mention two commands in a single line unless really required.

```
2. Install tools used by the build and test system.
```shell
make goget-tools
```

> 3. Build the executable in `cmd/odo`.

Reviewer 2: I understand here that the executable will be placed in `cmd/odo/`, which is not the case.

```suggestion
3. Build the executable.
```

or (but I'm not sure the non-developer user is interested in this level of details):

```suggestion
3. Build the executable from the sources in `cmd/odo`.
```

Reviewer: +1 for first one. I came here to add that feedback but found @feloy had already mentioned it.

Author: Taking the second suggestion into account, because I don't think any non-developer would be interested in building it from the source.

```shell
make bin
```
4. Install the executable in the system's GOPATH.
```shell
make install
```
5. Run the command to verify that it was installed properly.
```shell

> odo

Reviewer 2: Running `odo version` would help the user check that he doesn't have a previously installed one he completely forgot in another PATH's directory.

Reviewer: To run the odo binary built from the source, the user would have to do `./odo` and not `odo`. 
* `./odo` refers to the one built using steps mentioned here.
* `odo` refers to a previously installed binary that is already present in `PATH`.

Author: `./odo` would this also be true for windows?

Reviewer 2: > To run the odo binary built from the source, the user would have to do `./odo` and not `odo`.
> 
> * `./odo` refers to the one built using steps mentioned here.
> * `odo` refers to a previously installed binary that is already present in `PATH`.

But the step `make install` has installed `odo` in `$GOPATH/bin`, which is, probably, in the `PATH`

Reviewer: > But the step `make install` has installed `odo` in `$GOPATH/bin`, which is, probably, in the `PATH`

Ah. Totally missed that. I use `make && cp -f odo ~/bin/odo` on my local system so I was under that mindset.

Author: I am a little confused about this discussion. Does the current content look okay?

Reviewer: Yes.
